include/envoy/upstream/upstream.h

```diff
@@ -778,11 +778,29 @@ class ClusterInfo {
    */
   virtual ResourceManager& resourceManager(ResourcePriority priority) const PURE;
 
+  // Configuration to determine the transport socket factory to use.
+  struct TransportSocketFactoryOption {
+    // upstream networking address trying to connect to, for debugging purpose only.
+    const std::string address_;
+    // the metadata associated with upstream endpoint.
+    const envoy::api::v2::core::Metadata metadata_;
+  };
+
+  /**
+   * @return Network::ClusterInfo::TransportSocketFactoryOption the default option
+   *         to select a transport socket for a given cluster.
+   */
+  static TransportSocketFactoryOption DefaultTransportSocketFactoryOption() {
+    return TransportSocketFactoryOption{"", envoy::api::v2::core::Metadata()};
+  }
+
   /**
    * @return Network::TransportSocketFactory& the factory of transport socket to use when
    *         communicating with the cluster.
    */
-  virtual Network::TransportSocketFactory& transportSocketFactory() const PURE;
+  virtual Network::TransportSocketFactory&
+  transportSocketFactory(const TransportSocketFactoryOption& option =
```

Comment:
no default argument in virtual method. https://google.github.io/styleguide/cppguide.html#Default_Arguments

Reply:
Sure. but then we have two options

1. update all `transportSocketFactory()` callers to invoke this `defaulTransportSocketFactoryOption` as argumetn
1. create an overloading function with 0 argument, basically two `transportSocketFactory`, so existing invocations won't have to be modified.

which you think is better? second one obvious results in smaller changes though.

Reply:
hmm, speaking of that, I just realized we have two more call sites of the `ClusterInfo::transportSocketFactory`. 

1. [Router Filter](https://github.com/envoyproxy/envoy/blob/eff020170c6267e6c8dc235473f7fc85c5b1e07d/source/common/router/router.cc#L94)
2. [Dynamic Forward proxy filter](https://github.com/envoyproxy/envoy/blob/a659c5ecfd93a9d3a6580cbc968aa3fcca8eb480/source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc#L59)

Their purpose is only checking socket `implementsSecureTransport` though, but I  think to ensure features working all the cases, somehow we have to return the same type of socket factory as upstream_impl choose?

Reply:
either way works for default arguments, mildly prefer 1 with changing the argument to absl::optional so the default will be absl::nullopt.

for implementsSecureTransport I guess this should reflect the real upstream transport socket selection. Don't worrying about Dynamic Forward proxy filter, because that one requires dynamic forward cluster, so EDS is not in consideration. For Router Filter, I guess we already know which transport socket factory will be used when transportSocketFactory is called?

Reply:
`Router::FilterUtility::setUpstreamScheme` as a static function is invoked in router filter.

My understanding of the router's role

1. downstream request processed by filter
1. router process request and route based on path/xxx to different cluters.
1. (maybe) load balancing decision is made after route? what's the name in envoy component?
1. http connection pool is being selected.
1. connection is being established (`upstream_impl.cc`)

So by the time router is requesting the resolution of the `transportSocketFactory`, it does not know which upstream endpoint is connected to... 

However, I looked at the usage, it merely look at `transportSocketFactory().implementsSecureTransport()`. based on this decision, post processed the url (remove port 80 or 443...)

So seems like this won't affect connection with upstream? And to be consistent, can we just add an `ClusterInfo::implementSecureTransport`, such that it returns true if all the socket types return `implementSecureTransport` true?

This is consistent with our [discussion in the doc](https://docs.google.com/document/d/1lpshO7E-2KQ5SC519oiRe25TPAXp1HUcyVzZ4QM4IeU/edit?disco=AAAADb9FJUk)

Reply:
It does affect, I believe upstream HTTP implementations will reject the request with incorrect scheme (i.e. sending :scheme: http to HTTPS endpoint or vice versa).

It seems like in router.cc, Router::FilterUtility::setUpstreamScheme is called at https://github.com/envoyproxy/envoy/blob/master/source/common/router/router.cc#L505, which after ConnPool selecting at https://github.com/envoyproxy/envoy/blob/master/source/common/router/router.cc#L453, so you should be able to know which host is selected there. It might need some sort of refactoring however.

Reply:
okay, now it's addressed.

1. router part conn pool is selected by connection is created yet, so has access to host metadata, which is used to resolve the transport socket factory as the same upstream is doing.
2. `health_checker_impl` is different, it creates connection, so having access to the connection/codecclient, but not the factory/host information, thus I end up adding `implementSecureTransports` in transport socket class.